class-based api

jeffzi opened the pull request:
As discussed in #253, I submit a prototype of a class-based API. I think the best is that @cosmicBboy creates a branch dedicated to this prototype.

- [x] implement `SchemaModel`, a similar concept to pydantic.BaseModel. 
- [x] implement annotations: `DataFrame[subtype of SchemaModel]`, `Series[dtype]`, `Index[dtype]`
- [x] implement `check_types` decorator 
- [x] implement `ColumnField`, `IndexField`, `MultiIndexField` concepts to pass simple built-in checks and other arguments exposed by `schema_components.Column`, etc. (not a big fan of the "Field" suffix).
- [x] implement checks as decorators (similar to pydantic validators)
- [x] implement schema level attributes such as transformer, coerce, strict, etc.
- [ ] documentation

Simply put, the end-user creates a subclass of `SchemaModel` which must have class attributes annotated. Non-annotated attributes are ignored. Supported annotations at the class level are `Series` and `Index` at the moment, others will raise an error.

Behind the scene, `SchemaModel` can be converted to a regular `DataFrameSchema` with its class method `get_schema()`. The `check_types()` decorator inspects the function's annotations, transform them into `DataFrameSchema` and validate them.

```python
import pandera as pa
from pandera import Column, DataFrameSchema, check_types
from pandera.typing import DataFrame, Index, SchemaModel, Series

class A(SchemaModel):
    a: Series["int"]
    b: Series["string"]
    idx: Index["string"]

expected = DataFrameSchema(
    columns={"a": Column("int"), "b": Column("string")}, index=pa.Index("string")
)

assert expected == A.get_schema()

class B(SchemaModel):
    b: Series["int"]

# will raise a SchemaError
@check_types()
def to_b(df: DataFrame[A]) -> DataFrame[B]:
    return df
```

Some notes:
* Initially, we thought of `BaseSchema` but I think `SchemaModel` conveys the concept better.
* I did not consider `SeriesSchema` for now.
* I implemented simple testing but rigorous testing will be required.
* We need to design an annotation for MultiIndex. The most natural would be multiple generics like `MultiIndex["int", "int"]`.
* I really like the fact that it's just a translation layer on top of pandera.

When the base API for annotation is satisfactory, next would be `ColumnField`?

cosmicBboy commented:
Thanks for your contribution @jeffzi! This looks 🔥🔥🔥

I'll get to review this and also try it out locally over the weekend.

jeffzi commented:
Some technical notes:
1. The `typing` module is still maturing. I've opted to leverage the small lib [typing_inspect](https://github.com/ilevkivskyi/typing_inspect) written by a core typing contributor. It's easier for prototyping and supporting multiple python versions. I poked at pydantic and [typer](https://typer.tiangolo.com/) code and saw that they reimplemented part of `typing_inspect`. It's something to have in mind if you decide this API can be part of pandera :)
2. mypy raises an error on `pandera.typing.Series['int']`  -> Value of type variable "Dtype" of "Series" cannot be "int" but not on `pandera.typing.Series['string']`. I'm guessing mypy evaluates 'int' to the type int.  ¯\_(ツ)_/¯
3. `pandera.typing.Series[pandera.Int]` fails at runtime but not  `pandera.typing.Series[pandera.Int.value]`

FYI I'll be on holiday from Aug 15 to Aug 24, probably not very active until I come back :)

cosmicBboy commented:
Cool, thanks for the notes @jeffzi 

I made a [typing](https://github.com/pandera-dev/pandera/tree/typing) branch for this prototype, I guess you can make a new PR against this branch?

A few points to keep track of here:
- looks like `typing_inspect` supports 3.7.4+, is there some sort of backport for 3.6? Note sure if [typing-extensions](https://pypi.org/project/typing-extensions/) will get us compatibility with typing_inspect. If it does then dropping 3.5 support #142 is not a blocker for this functionality.
- If we absolutely need to we could re-implement what we need of the `typing_inspect` package but I'm generally not one for re-inventing the wheel. 3.5 maintenance ends 2020-09-13, but 3.6 doesn't end until the end of 2021 according to [this](https://devguide.python.org/#status-of-python-branches).

> I really like the fact that it's just a translation layer on top of pandera.

same 👍

> I did not consider SeriesSchema for now.

This is fine, I'm not sure how widely used this schema class is, anyway. Can come later.

> (not a big fan of the "Field" suffix)

It's not ideal, but I was thinking that there needs to be a meaningful to not confuse users between the `pandera.Column`  and `pandera.typing.Column` classes, since their APIs are likely going to be different, hence the `Field` suffix... perhaps it's enough that they're in different submodules, what are your thoughts? Would be open to dropping the suffix.

jeffzi commented:
> I made a class-api branch for this prototype, I guess you can make a new PR against this branch?
I switched the base branch for this PR to the `class-api` branch you created. Let me know if that's enough.

> looks like typing_inspect supports 3.7.4+
Actually it needs the  [typing](https://pypi.org/project/typing/) backport for python < 3.5. 
https://github.com/ilevkivskyi/typing_inspect/blob/0950ff8bf27472add2ab637bbf700e0e01494e2b/setup.py#L37
Otherwise, I agree re-inventing the wheel is unnecessary, especially at this stage.

> mypy raises an error on pandera.typing.Series['int'] -> Value of type variable "Dtype" of "Series" cannot be "int" but not on pandera.typing.Series['string']. I'm guessing mypy evaluates 'int' to the type int. ¯_(ツ)_/¯
Thanks to the support of built-in types, I could try `pandera.typing.Series[int]`. It works if I add `int` to the definition of Dtype TypeVar [here]( https://github.com/jeffzi/pandera/blob/f4d5bdb80c187c62f9e4057a504f39199ec0f7c6/pandera/typing.py#L14).

> pandera.typing.Series[pandera.Int] fails at runtime but not pandera.typing.Series[pandera.Int.value]
We can have aliases to make it less ugly:
```python
from pandera.typing import Series

Int64 = PandasDtype.Int64.value
Series[Int64]
```
In the future, it would be even nice if pandera supported numpy dtype `pandera.typing.Series[np.int64]`, `pandera.Column(np.int64)`. 

We can move on to `ColumnField` if you don't see any blockers.

cosmicBboy commented:
> I switched the base branch for this PR to the class-api branch you created. Let me know if that's enough.

Excellent! thanks ✨

> We can move on to `ColumnField` if you don't see any blockers.

Yes I think we can move forward with `ColumnField`. Also, feel free to rename to `typing.Column` if you think it's more elegant/clear, still not sure if it's confusing given the `pandera.Column` class.

> In the future, it would be nice if pandera supported numpy dtypes: pandera.typing.Series[np.int64]

Yes! working on this #262 next when I have time in the next few weeks :)

jeffzi commented:
I've implemented the core functionalities. I'm going to break down the discussion into several comments.
## Basic usage

```python
import pandera as pa
from pandera.typing import Index, Series


class BaseSchema(pa.SchemaModel):

    year: Series[str] = pa.Field(le=2000, coerce=True)
    month: Series[str]

    # Define schema-wide options, i.e. DataFrameSchema(..., strict, coerce, name, etc.)
    class Config:
        name = "BaseSchema"
        strict = True
        coerce = True
        foo = "bar"  # not a valid option, ignored


class FinalSchema(BaseSchema):
    year: Series[int]  # overwrite base dtype
    passengers: Series[int]
    idx: Index[int] = pa.Field(ge=1950)

    class Config:
        name = "FinalSchema"  # overwrite base option
        strict = False  # overwrite base option


expected_schema = pa.DataFrameSchema(
    columns={
        "year": pa.Column(
            int,
            checks=pa.Check.less_than_or_equal_to(2000),
            coerce=True,
        ),
        "month": pa.Column(str),
        "passengers": pa.Column(int),
    },
    index=pa.Index(int, pa.Check.greater_than_or_equal_to(1950)),
    coerce=True,
    strict=False,
    name="FinalSchema",
)

actual_schema = FinalSchema.to_schema()
assert expected_schema == actual_schema

print(f"{actual_schema.columns['passengers'].name=}")
#> actual_schema.columns['passengers'].name='passengers'
print(f"{(actual_schema.index.name is None)=}")
#> (actual_schema.index.name is None)=True
```

<sup>Created on 2020-09-15 by the [reprexpy package](https://github.com/crew102/reprexpy)</sup>

* `pa.Field` allows to defines properties on Column **or** Index, as well as built-in checks
* Columns are automatically named, a singular index is not. (more about MultiIndex later).
* Schema-wide options are defined in an inner class, exactly [like pydantic](https://pydantic-docs.helpmanual.io/usage/model_config/).
* Inheritance works as expected. Depending on the situation, it should be easier to reuse schemas with the class-api. The limitation is that you can only append columns, indices, and checks in the subclass, not remove them. Annotations and Config options can be replaced. Users can easily switch to regular API via `to_schema()` if needed.
* Users can create their own reusable `Field` by subclassing [pandera.schema_model.FieldInfo](https://github.com/jeffzi/pandera/blob/c4d556cf40d6d78a7be74ba4e8d94fc76b002835/pandera/schema_model.py#L52).

jeffzi commented:
## Typing

The API supports python >= 3.6 because variable annotations were introduced in 3.6 https://www.python.org/dev/peps/pep-0526/

In order to support [PEP 563 -- Postponed Evaluation of Annotations](https://www.python.org/dev/peps/pep-0563), the dtype must be resolvable. For example, `Series["str"]` works but not `Series["string"]` because the type `string` cannot be resolved. PEP 563 will be the default in python 3.9 or with `from __future__ import annotations` (available with >= python 3.7).

Series[pd.StringDtype()]` doesn't work either because it's an instance and not a type. The best solution would be to support numpy dtypes (adressed by #262 ) and [pandas dtypes](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.CategoricalDtype.html#pandas.CategoricalDtype) as well.

jeffzi commented:
## MultiIndex
```python
import pandera as pa
from pandera.typing import Index, Series

class MultiIndexSchema(pa.SchemaModel):
    
    year: Index[int]
    month: Index[int]
    passengers: Series[int]

    class Config:
        multiindex_name = "time"
        multiindex_strict = True
        multiindex_coerce = True
        
print(MultiIndexSchema.to_schema().index)
#> MultiIndex(
#>     columns={
#>         "year": "<Schema Column: 'year' type=<class 'int'>>",
#>         "month": "<Schema Column: 'month' type=<class 'int'>>"
#>     },
#>     checks=[],
#>     index=None,
#>     transformer=None,
#>     coerce=True,
#>     strict=True
#> )
```

<sup>Created on 2020-09-15 by the [reprexpy package](https://github.com/crew102/reprexpy)</sup>

jeffzi commented:
## Validators

### Column / Index checks

```python
class ValidatorSchema(pa.SchemaModel):
    a: Series[int] = pa.Field(gt=0, coerce=True)
    abc: Series[int]
    idx: Index[str]
    
    # check kwargs are allowed
    @pa.validator("^a", regex=True, name="foobar") 
    def check_prefix(a: Series[int]) -> Series[bool]:
        return a > 0

    @pa.validator("idx") 
    def check_(idx: Index[int]) -> Series[bool]:
        return idx.str.contains("dog")
```

### DataFrame validator / transformer

```python
import pandas as pd
import pandera as pa
from pandera.typing import Index, Series

class A(pa.SchemaModel):
    a: Series[int]

    @pa.dataframe_validator
    def value_lt_100(df: pd.DataFrame) -> bool:
        return df < 100
    
    @pa.dataframe_transformer
    def neg_a(df: pd.DataFrame) -> pd.DataFrame:
        df["a"] = -df["a"]
        return df
    
class B(A):
    
    @pa.dataframe_transformer
    def transform_abs(df: pd.DataFrame) -> pd.DataFrame:
        return df.abs()
    
B.to_schema().transformer # forbidden to have more than 1 transformer
#> Traceback (most recent call last):
#> <ipython-input-8-8bb76cfd242b> in <module>
#> ----> 1 B.to_schema().transformer # forbidden to have more than 1 transformer
#> ~/Projects/development/pandera/pandera/schema_model.py in to_schema(cls)
#>     240             tr = _extract_transformer(fn)
#>     241             if transformer and tr:
#> --> 242                 raise SchemaInitError(
#>     243                     f"{cls.__name__} can only have one 'dataframe_transformer'."
#>     244                 )
#> SchemaInitError: B can only have one 'dataframe_transformer'.
```

<sup>Created on 2020-09-15 by the [reprexpy package](https://github.com/crew102/reprexpy)</sup>

* Pydantic validators combine checks and transformers. It'd be better to find a more suitable name to not confuse people coming from pydantic. `@check`, `@dataframe_check`?

* For now, you cannot overwrite a transformer. I'm thinking of having a warning and exposing `get_transformer()` or a `__transformer__` class attribute to allow the subclass to pipe the transformer of its parent.
* Validator are not methods. Pydantic internally wraps them with `classmethod`. In our case, I don't see the added value of having access to `cls` during checks/transformations. On the other hand, linters complain `Method should have 'self' as first argument`. Alternatively, We could wrap them with `staticmethod`.

jeffzi commented:
## Open questions

* Implement field aliases like pydantic?
* Rename `@validator`, `@dataframe_validator` -> `@check`, `@dataframe_check` ?
* What to do with subclass re-declaring a dataframe transformer?
* `static` or `classmethod` for model decorators?
* open a new issue for pandas dtype

Sorry for the very long explanations. I hope it will spark interesting discussions :)

cosmicBboy commented:
thanks @jeffzi for this herculean effort! just did a quick skim, and overall looks fantastic. I'll have more comments/questions later, but on the more immediate side, looks like we'll need to do a few things on the CI:

- I'll be dropping support for python3.5 in the next release since it's been EOL, so we can ignore failures on those tests
- looks like pylint is complaining for 3.6+ tests. We can configure pylint to ignore many of those (some of them are docstring-related, which we should add as the last documentation piece)

> Implement field aliases like pydantic?

Are you referring to [this feature](https://pydantic-docs.helpmanual.io/usage/model_config/#alias-generator)? I think it would be a nice-to-have, perhaps something to implement in a future PR, but I don't see at as necessary for this core functionality.

> Rename @validator, @dataframe_validator -> @check, @dataframe_check ?

Interesting, so I guess `pydantic` is technically a parser (where data validation is guaranteed via parsing). I've actually been re-thinking the `transformer` functionality in pandera. #252 is one such proposal, but I can imagine other ways of separating the concerns of "parsing" and "validation". In a way, the pydantic nomenclature for `validators` may be slightly misleading in that it's actually parsing data and returning clean values, where the constrained types (and constraints in the `Field` objects) are conceptually similar to pandera `Check`s.

To answer your question though, I think `@check` and `@dataframe_check` are good names to fit with the pandera naming convention.

> What to do with subclass re-declaring a dataframe transformer?

I kind of see `transformers` in pandera as a vestigial feature that I slapped on there in the early days of the package, and I don't think it's a super-well-thought-out feature, since it's literally pandas code that one can pipe after validating a dataframe `schema(df).pipe(...)`. I would be okay with not including this functionality in the class-based API (and actually deprecating `transformer` altogether in pandera) and then adding parsing functionality once there's a more principled approach to it in the pandera API.

> static or classmethod for model decorators?

What are the reasons to use the `cls` arg in pydantic? I think `staticmethod` is fine to use here.

> open a new issue for pandas dtype

what needs to be done here?

jeffzi commented:
> thanks @jeffzi for this herculean effort! 

You're very welcome!

> Are you referring to this feature?

Yes, specifically `Field(..., alias=<alias>)`. Definitely something for a second pass.

> Interesting, so I guess pydantic is technically a parser ...

I agree pydantic validators are misleading.  I'd rather have a clean separation between checks/validators and transformers/converters, which is the approach taken by [attrs](https://github.com/python-attrs/attrs). 

Pydantic's constrained types can actually do anything a Field can do. See the example [here](https://pydantic-docs.helpmanual.io/usage/types/#custom-data-types). Anyway, I don't think constrained types are desirable for pandera. As I mentioned, end-users can subclass `FieldInfo` which contains metadata about column/index (checks, coerce, etc.).  `Field`  is a function that returns `Any` to make the linters happy and have a user-friendly interface. Linters wouldn't like us to assign a `FieldInfo` to a `Series[int]` :)

Parsing does sound like a natural path for pandera. I'll remove transformer until the core api has matured.

> What are the reasons to use the cls arg in pydantic?

After thinking about it, you can access [model properties](https://pydantic-docs.helpmanual.io/usage/models/#model-properties) or custom methods. For example, you could imagine someone adding a new attribute to `Config` and access it in a `check`. Basically I made a u-turn and would prefer a `classmethod` for power users.

> open a new issue for pandas dtype

`DataFrameSchema({"column1": Column(pd.CategoricalDtype)})` raises `TypeError: python type '<class 'pandas.core.dtypes.dtypes.CategoricalDtype'>' not recognized as pandas data type`. 
`Series` is a generic type and must receive a type parameter: `Series[pd.CategoricalDtype]`. If [SeriesSchemaBase.coerce_dtype ](https://github.com/pandera-dev/pandera/blob/b7936d5b5ac11f14bb0894014a24c6da8f94c4dc/pandera/schemas.py#L746) could accept types on top of instances it would solve the typing issue nicely. I can also add that to this PR if it's fine with you.

cosmicBboy commented:
> If SeriesSchemaBase.coerce_dtype could accept types on top of instances it would solve the typing issue nicely. I can also add that to this PR if it's fine with you.

go ahead!

jeffzi commented:
I made the following changes:
- Fixed pylint and mypy errors. Remaining failures are due to py3.5 that will be dropped + increased coverage to 100% for class api.
- Renamed validators to `@check` and `@dataframe_check`
- checks decorators are automatically decorated with `classmethod`. Note that mypy will complain, same as pydantic, without the explicit `@classmethod`. 

```python
    class Schema(pa.SchemaModel):
        a: Series[int]

        @pa.check("a")
        def int_column_lt_100(cls, series: pd.Series) -> Iterable[bool]:
            assert cls is Schema
            return series < 100
``` 
- Dropped `@dataframe_transformer`
- Added support for dtype as strings via `Literal` annotations. I added [aliases for supported PandasDtypes](https://github.com/jeffzi/pandera/blob/98d3d7a656ccffd12eed0518a220d7c8777868c8/pandera/typing.py#L115) to reduce boilerplate. `from pa.typing import Category, String, etc.`

```python
class SchemaCategory(pa.SchemaModel):
    literal_cat: Series[Literal["category"]]
    alias_cat: Series[pa.typing.Category]
``` 
- Added support for Pandas ExtensionDtype classes as `pandas_dtype` argument for `Column` and `Index`. An [informative error](https://github.com/jeffzi/pandera/blob/98d3d7a656ccffd12eed0518a220d7c8777868c8/pandera/schemas.py#L768) is raised if the class has mandatory arguments and cannot be instantiated. e.g: [DatetimeTZDtype](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DatetimeTZDtype.html#pandas.DatetimeTZDtype) requires the `tz` argument, `Column(pd.DatetimeTZDtype)` will raise an error.

```python
import pandas as pd
import pandera as pa
from pandera.typing import Series

class Schema(pa.SchemaModel):
    col: Series[pd.StringDtype]

assert Schema.to_schema().columns["col"].dtype == pa.Column(pd.StringDtype).dtype
``` 

With the last 2 points, the class-api should be instinctive to use :sparkles:

cosmicBboy commented:
@jeffzi tests are passing now after #272, there are some code coverage drops in  `dtypes.py`, `schemas.py`, and `decorators.py`: https://codecov.io/gh/pandera-dev/pandera/compare/91e21a2bf769096e80cf7c08c1311ee06aadebb6...d4f5d4563c52ad6693f6ea6a12c6f8541121d5b7/tree?path=pandera

I think some of these are technical debt pre-dating this PR, let me know if you need help identifying which pieces those are.

One major piece that's missing in CI is to support pre-1.0.0 pandas to make sure legacy-pandas functionality is working. I can work on this in the different PR.

cosmicBboy commented on pandera/schemas.py:
```diff
@@ -32,6 +32,19 @@
 
 PandasDtypeInputTypes = Union[str, type, PandasDtype, PandasExtensionType]
 
+if pd.__version__ < "1.0.0":
```
might be more robust to use `version.packaging` as in [here](https://github.com/pandera-dev/pandera/blob/master/tests/test_dtypes.py#L21) and [here](https://github.com/pandera-dev/pandera/blob/master/tests/test_dtypes.py#L120-L123) as opposed to relying on string comparison

cosmicBboy commented on tests/test_typing.py:
```diff
@@ -0,0 +1,280 @@
+from pandera.model import SchemaModel
+from pandera.typing import Series
+
+LEGACY_PANDAS = pd.__version__ < "1.0.0"
```
same here re: using `version.packaging`

jeffzi replied:
done :+1:

jeffzi commented:
I verified that all tests pass with pandas 0.25.3 and 1.1.2. We shouldn't have to modify anything once #276 is ready.

jeffzi commented:
I increased the coverage. The remaining drop in the `schemas.py` module will be fixed with #276

jeffzi commented:
I increased the coverage. `dtypes.py` will be at 100% and schemas.py will be fixed with #276.

I've also added a convenience method `SchemaModel.validate()` to reduce the boilerplate for the common operation `SchemaModel.to_schema().validate()`. 

Regarding documentation, I think it would be more clear to have another PR once this one is accepted.

cosmicBboy commented on pandera/model_components.py:
```diff
@@ -0,0 +1,273 @@
+def Field(
+    *,
+    eq: Any = None,
+    neq: Any = None,
```
what do you think of renaming this to `ne`? that would be consistent with the [python operator](https://docs.python.org/3/library/operator.html#operator.ne) and [pandas function](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.ne.html).

(fyi I also made alias methods for these in `Check` for convenience #269)

jeffzi replied:
Makes sense. I committed the change :+1:

cosmicBboy reviewed (commented):
looks good to me! let me know what you think about renaming `neq` to `ne` (see inline comment). Once we discuss that this should be good to go.

jeffzi commented:
@cosmicBboy Thanks. I've cleaned up minor stuff. Good to go on my side too.

cosmicBboy reviewed (approved):
✅ thanks @jeffzi this is awesome, thanks for contributing!

also, let me know if you need any help with documentation
